src/stream/src/task/barrier_manager.rs

```diff
@@ -268,97 +276,111 @@ pub(super) struct LocalBarrierWorker {
-
-    actor_failure_rx: UnboundedReceiver<(ActorId, StreamError)>,
 }
 
 impl LocalBarrierWorker {
     pub(super) fn new(
         actor_manager: Arc<StreamActorManager>,
-        initial_partial_graphs: Vec<InitialPartialGraph>,
+        initial_partial_graphs: Vec<DatabaseInitialPartialGraph>,
     ) -> Self {
-        let (event_tx, event_rx) = unbounded_channel();
-        let (failure_tx, failure_rx) = unbounded_channel();
-        let shared_context = Arc::new(SharedContext::new(
-            &actor_manager.env,
-            LocalBarrierManager {
-                barrier_event_sender: event_tx,
-                actor_failure_sender: failure_tx,
-            },
-        ));
+        let state = ManagedBarrierState::new(actor_manager.clone(), initial_partial_graphs);
         Self {
-            state: ManagedBarrierState::new(
-                actor_manager.clone(),
-                shared_context.clone(),
-                initial_partial_graphs,
-            ),
+            state,
             await_epoch_completed_futures: Default::default(),
             control_stream_handle: ControlStreamHandle::empty(),
             actor_manager,
-            current_shared_context: shared_context,
-            barrier_event_rx: event_rx,
-            actor_failure_rx: failure_rx,
         }
     }
 
     fn to_debug_info(&self) -> LocalBarrierWorkerDebugInfo<'_> {
         LocalBarrierWorkerDebugInfo {
-            running_actors: self.state.actor_states.keys().cloned().collect(),
-            managed_barrier_state: self.state.to_debug_info(),
+            managed_barrier_state: self
+                .state
+                .databases
+                .iter()
+                .map(|(database_id, state)| (*database_id, state.to_debug_info()))
+                .collect(),
             has_control_stream_connected: self.control_stream_handle.connected(),
         }
     }
 
+    pub(crate) fn get_or_insert_database_shared_context<'a>(
+        current_shared_context: &'a mut HashMap<DatabaseId, Arc<SharedContext>>,
+        database_id: DatabaseId,
+        actor_manager: &StreamActorManager,
+    ) -> &'a Arc<SharedContext> {
+        current_shared_context
+            .entry(database_id)
+            .or_insert_with(|| Arc::new(SharedContext::new(database_id, &actor_manager.env)))
+    }
+
+    async fn next_completed_epoch(
+        futures: &mut HashMap<DatabaseId, FuturesOrdered<AwaitEpochCompletedFuture>>,
+    ) -> (
+        DatabaseId,
+        PartialGraphId,
+        Barrier,
+        StreamResult<BarrierCompleteResult>,
+    ) {
+        poll_fn(|cx| {
+            for (database_id, futures) in &mut *futures {
+                if let Poll::Ready(Some((partial_graph_id, barrier, result))) =
+                    futures.poll_next_unpin(cx)
+                {
+                    return Poll::Ready((*database_id, partial_graph_id, barrier, result));
+                }
+            }
+            Poll::Pending
```

Comment:
If there are a considerable amount of databases, we may consider constructing them into an outer `FuturesUnordered`.

Reply:
Might be difficult to do so. 

If we create a temporary `FuturesUnordered` in every select, it may hurt performance somehow since we need to allocate new memory to create new `FuturesUnordered`. If we store the `FuturesUnordered` as a field of local barrier worker, the future has to be static and otherwise we will need to self-reference the `DatabaseManagedBarrierState` stored in local barrier worker. If we make the future to be static, we need to move the ownership of `DatabaseManagedBarrierState` to the future, and we will not be able to access the `DatabaseManagedBarrierState` before the future returns.